src/posit/connect/users.py

```diff
@@ -10,21 +10,53 @@
 
 from .config import Config
 from .paginator import _MAX_PAGE_SIZE, Paginator
-from .resources import Resources
-
-
-class User(TypedDict, total=False):
-    guid: str
-    email: str
-    username: str
-    first_name: str
-    last_name: str
-    user_role: str
-    created_time: datetime
-    updated_time: datetime
-    active_time: datetime
-    confirmed: bool
-    locked: bool
+from .resources import Resource, Resources
+
+
+class User(Resource):
+    @property
+    def guid(self) -> Optional[str]:
+        return self.get("guid")
+
+    @property
+    def email(self) -> Optional[str]:
+        return self.get("email")
+
+    @property
+    def username(self) -> Optional[str]:
+        return self.get("username")
+
+    @property
+    def first_name(self) -> Optional[str]:
+        return self.get("first_name")
+
+    @property
+    def last_name(self) -> Optional[str]:
+        return self.get("last_name")
+
+    @property
+    def user_role(self) -> Optional[str]:
+        return self.get("user_role")
+
+    @property
+    def created_time(self) -> Optional[datetime]:
```

Comment:
I guess these are just hints for the IDE, they don't seem to enforce anything. This doesn't return datetime in the tests. 

So we should at least switch to `Optional[str]`. But if they're just informative for tab completion or whatever, perhaps we should just say what they're supposed to be, whether or not they're present in a given API response?

```suggestion
    def created_time(self) -> str:
```

Reply:
I like the expressive nature of matching the API response schema, but this breaks down between versions of Connect. Therefore, from a software correctness perspective, `Optional` is always correct. 

That logic defines the return type on this method as `Optional[str].`

Do you come to the same conclusion?

Reply:
Feels a little hyperbolic to think that we don't know what the API will respond in the future. I also figure we'd be adding backwards compat logic if we rename fields. So IMO I'd think we'd want a type hint that was precise and 99% accurate rather than something that was vague and 100% accurate. But then again, I'm probably not the person you want to take advice from on type annotations ;)

Reply:
That's fair. Apologies if the following is obvious...

I tend to think of the type system from a software correctness perspective. So, always returning Optional for these properties would ensure that software written with this version of the SDK will continue to run successfully when a field name changes. This puts the responsibility on the consumer to trigger an error if `created_time` returns None and their application requires a non-None value. Anyone analyzing their code with MyPy will be forced to consider this scenario if we return `Optional` here.

e.g.,
```
created_time = user.created_time
if created_time is None:
    raise RuntimeError

# continue using created_time
```

The flip side returns `str` and then raises an error if the value does not exist. This puts the responsibility on us to keep the SDK up to date and the user to make sure they are using the latest version of the SDK.

```
def created_time(self) -> str:
    if 'created_time' in self:
        return self['created_time']
    raise RuntimeError("'created_time' not found in the response object. Please upgrade to the latest version...)
```

--

After thinking it through, I realized that the second option is a better user experience overall. I want to modify the properties to mimic the expected API behavior and provide actionable error messages.

Reply:
IMO we shouldn't let fear of the possibility of changing field names in the API response drive too much of our decisionmaking since we're much more likely to add new fields than remove them, and we'd only remove after a deprecation cycle, which we'd be managing in the SDK. 

I think it's also safe to assume (because we can strongly recommend it) that interactive users are running on the latest version of the SDK, which will know about API changes. If we know that a field only was added in some new version of connect, we can check the Connect version in the `property` and error if it's not new enough. 

The place where we want to be forgiving of changes is with deployed content, we want it to be robust to Connect upgrades. Type hints are irrelevant there because the content isn't being edited. There is a risk of some distant future, after a deprecation cycle, we remove an attribute, and you could have content referencing it with an old SDK version. But, we could also provide a script (using the SDK) that checks package versions in your content before you upgrade Connect and alert you to which ones you need to upgrade the sdk in.